File: project_euler/Problem048/Problem048.js
function Problem048() {
    let target = 1000
    let product = Array(2)
    let auxiliary = Array(2)
    let result = 0
    for (let i = Math.floor(target / 10) * 10 + 1; i <= target; ++i) {
        getPower(i, i, product, auxiliary)
        result += product[1] * 1e5 + product[0]
    }
    for (let i = Math.floor(target / 10) - 1; i >= 0; --i) {
        for (let j = 1; j < 10; ++j) {
            let value = i * 10 + j
            getPower(value, value, product, auxiliary)
            result += product[1] * 1e5 + product[0]
        }
    }
    console.log(result % 1e10)
}

function getPower(base, index, product, auxiliary) {
    if (1 === index) {
        product[0] = base
        product[1] = 0
        return
    }

    getPower(base, index >>> 1, product, auxiliary)
    getSquare(product, auxiliary)
    if (0 !== (index & 1)) {
        getProduct(base, product, auxiliary)
    }
}

function getSquare(product, auxiliary) {
    auxiliary[0] = product[0] * product[0]
    auxiliary[1] = product[0] * product[1] * 2
    carry(product, auxiliary)
}

function getProduct(value, product, auxiliary) {
    auxiliary[0] = product[0] * value
    auxiliary[1] = product[1] * value
    carry(product, auxiliary)
}

function carry(product, auxiliary) {
    if (auxiliary[0] >= 1e5) {
        auxiliary[1] += Math.floor(auxiliary[0] / 1e5)
        auxiliary[0] %= 1e5
    }
    auxiliary[1] %= 1e5
    product[0] = auxiliary[0]
    product[1] = auxiliary[1]
}

module.exports = {
    Problem048 : Problem048
};

File: project_euler/Problem048/Problem048.cpp
#include <iostream>
#include <vector>
using namespace std;

class Problem048 {
public:
    void solution() {
        int target = 1000;
        vector<int> product(2);
        vector<long long> auxiliary(2);
        long long result = 0L;
        for (int i = target / 10 * 10 + 1; i <= target; ++i) {
            get_power(i, i, product, auxiliary);
            result += (long long) product[1] * (long long) 1e5 + product[0];
        }
        for (int i = target / 10 - 1; i >= 0; --i) {
            for (int j = 1; j < 10; ++j) {
                int value = i * 10 + j;
                get_power(value, value, product, auxiliary);
                result += (long long) product[1] * (long long) 1e5 + product[0];
            }
        }
        cout << result % (long long) 1e10 << endl;
    }

private:
    void get_power(int base, int index, vector<int> &product, vector<long long> &auxiliary) {
        if (1 == index) {
            product[0] = base;
            product[1] = 0;
            return;
        }

        get_power(base, index >> 1, product, auxiliary);
        get_square(product, auxiliary);
        if (index & 1) {
            get_product(base, product);
        }
    }

    static void get_square(vector<int> &product, vector<long long> &auxiliary) {
        auxiliary[0] = (long long) product[0] * product[0];
        auxiliary[1] = (long long) product[0] * product[1] << 1;
        if (auxiliary[0] >= (long long) 1e5) {
            auxiliary[1] += auxiliary[0] / (long long) 1e5;
            auxiliary[0] %= (long long) 1e5;
        }
        auxiliary[1] %= (long long) 1e5;
        for (int i = 0; i < 2; ++i) {
            product[i] = (int) auxiliary[i];
        }
    }

    static void get_product(int value, vector<int> &product) {
        product[0] *= value;
        product[1] *= value;
        if (product[0] >= (int) 1e5) {
            product[1] += product[0] / (int) 1e5;
            product[0] %= (int) 1e5;
        }
        product[1] %= (int) 1e5;
    }
};
